aptos-move/framework/aptos-framework/sources/fee_distribution.move

```diff
@@ -0,0 +1,204 @@
+module aptos_framework::fee_destribution {
+    use std::error;
+    use std::option::{Self, Option};
+
+    use aptos_framework::aptos_coin::AptosCoin;
+    use aptos_framework::coin::{Self, AggregatorCoin};
+    use aptos_framework::system_addresses;
+    use aptos_framework::transaction_fee;
+
+    friend aptos_framework::transaction_validation;
+
+    /// When struct holding distribution ifnormation already exists.
+    const EDISTRIBUTION_INFO_EXISTS: u64 = 1;
+
+    /// Resource which holds the collected transaction fees and their receiver.
+    struct DistributionInfo has key {
+        balance: AggregatorCoin<AptosCoin>,
+        receiver: Option<address>,
+    }
+
+    /// Initializes the resource holding information for gas fees distribution.
+    /// Should be called by on-chain governance.
+    public fun initialize_distribution_info(aptos_framework: &signer) {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<DistributionInfo>(@aptos_framework),
+            error::already_exists(EDISTRIBUTION_INFO_EXISTS)
+        );
+
+        let zero = coin::initialize_aggregator_coin(aptos_framework);
+        let info = DistributionInfo {
+            balance: zero,
+            receiver: option::none(),
+        };
+        move_to(aptos_framework, info);
+    }
+
+    /// Called by transaction epilogue to collect the gas fees from the specified account.
+    public(friend) fun collect_fee(account: address, fee: u64) acquires DistributionInfo {
+        let distribution_info = borrow_global_mut<DistributionInfo>(@aptos_framework);
+        let dst_coin = &mut distribution_info.balance;
+        coin::collect_from(account, fee, dst_coin);
+    }
+
+    /// Sets the receiver of the collected fees for the next block.
+    public fun set_receiver(vm: &signer, receiver_addr: address) acquires DistributionInfo {
+        // Can only be called by the VM.
+        system_addresses::assert_vm(vm);
+        let distribution_info = borrow_global_mut<DistributionInfo>(@aptos_framework);
+        let _ = option::swap_or_fill(&mut distribution_info.receiver, receiver_addr);
+    }
+
+    /// Distributes collected transaction fees to the receiver. Should be called
+    /// at the beginning of each block.
+    public fun maybe_distribute_fees(vm: &signer) acquires DistributionInfo {
+        // Can only be called by the VM.
+        system_addresses::assert_vm(vm);
+        let distribution_info = borrow_global_mut<DistributionInfo>(@aptos_framework);
```

Comment:
we need to play safe to check that distribution_info is initialized otherwise fallback to previous behavior

Reply:
I think this has been addressed, currently we have the following: get info, if no coins (i.e. uninitialized), return, if there are coins but info has no proposer, we fallback and burn.